src/Models/HydrostaticFreeSurfaceModels/hydrostatic_free_surface_ab2_step.jl

```diff
@@ -0,0 +1,77 @@
+using Oceananigans.TimeSteppers: ab2_step_field!
+
+import Oceananigans.TimeSteppers: ab2_step!
+
+function ab2_step!(model::HydrostaticFreeSurfaceModel, Δt, χ)
+
+    workgroup, worksize = work_layout(model.grid, :xyz)
+
+    barrier = Event(device(model.architecture))
+
+    step_field_kernel! = ab2_step_field!(device(model.architecture), workgroup, worksize)
+
+    # Launch velocity update kernels
+
+    velocities_events = []
+
+    for name in (:u, :v)
+        Gⁿ = model.timestepper.Gⁿ[name]
+        G⁻ = model.timestepper.G⁻[name]
+        velocity_field = model.velocities[name]
+
+        event = step_field_kernel!(velocity_field, Δt, χ, Gⁿ, G⁻,
+                                   dependencies=Event(device(model.architecture)))
+
+        push!(velocities_events, event)
+    end
+
+    # Launch tracer update kernels
+
+    tracer_events = []
+
+    for name in tracernames(model.tracers)
+        Gⁿ = model.timestepper.Gⁿ[name]
+        G⁻ = model.timestepper.G⁻[name]
+        tracer_field = model.tracers[name]
+
+        event = step_field_kernel!(tracer_field, Δt, χ, Gⁿ, G⁻,
+                                   dependencies=Event(device(model.architecture)))
+
+        push!(tracer_events, event)
+    end
+
+    velocities_update = MultiEvent(Tuple(velocities_events))
+
+    # Update the free surface if not using a rigid lid once the velocities have finished updating.
+    free_surface_event = ab2_step_free_surface!(model.free_surface, velocities_update, model, χ, Δt)
```

Comment:
@francispoulin you wrote:

> I see you have an AB2 time stepper for this model. Is there a reason why the standard time steppers don't work? We tried to generalize them to work for any model but maybe didn't generalize them enough?

Our design doesn't generalize to the case of an implicitly-treated free surface. I do hope and think we can implement a general design that works for all models (and thereby eliminate this custom implementation for `HydrostaticFreeSurfaceModel`. For now we are taking a shortcut because we need the implementation soon, and don't have time to work on a general design (sometimes its also hard to come up with a general design before you know exactly what you need to do).

Reply:
> I see that you added free_surface as an argument in HydrostaticFreeSurfaceTendencyFields. Is this what is different and prevents you from using the time-stepping schemes?

The other issue is that the free surface is two-dimensional, and thus requires a different kernel launch configuration than the three-dimensional fields. We need to come up with abstractions that allow this process to be generalized for `prognostic_fields(model)`. It'd be really nice to have more generic code (and less code).

Reply:
Very helpful, thanks